pkg/square/builder.go

```diff
@@ -0,0 +1,238 @@
+		if err := blobWriter.Write(element.blob); err != nil {
+			return nil, fmt.Errorf("writing blob into sparse shares: %w", err)
+		}
+		// increment the cursor by the size of the blob
+		cursor += element.shareSize
+		endOfLastBlob = cursor
+	}
+
+	// write all the pay for blob transactions into compact shares. We need to do this after allocating the blobs to their
+	// appropriate shares as the starting index of each blob needs to be included in the PFB transaction
+	pfbTxWriter := shares.NewCompactShareSplitter(namespace.PayForBlobNamespace, appconsts.ShareVersionZero)
+	for _, iw := range c.pfbs {
+		iwBytes, err := iw.Marshal()
+		if err != nil {
+			return nil, fmt.Errorf("marshaling pay for blob tx: %w", err)
+		}
+		if err := pfbTxWriter.WriteTx(iwBytes); err != nil {
+			return nil, fmt.Errorf("writing pay for blob tx into compact shares: %w", err)
+		}
+	}
+
+	// defensively check that the counter is always greater in share count than the pfbTxWriter.
+	if c.pfbCounter.Size() < pfbTxWriter.Count() {
+		return nil, fmt.Errorf("pfbCounter.Size() < pfbTxWriter.Count(): %d < %d", c.pfbCounter.Size(), pfbTxWriter.Count())
+	}
+
+	// Write out the square
+	return WriteSquare(txWriter, pfbTxWriter, blobWriter, nonReservedStart, ss)
+}
+
+func (c *Builder) canFit(shareNum int) bool {
+	return c.currentSize+shareNum <= c.maxCapacity
+}
+
+func (c *Builder) isEmpty() bool {
+	return c.txCounter.Size() == 0 && c.pfbCounter.Size() == 0
+}
+
+type element struct {
+	blob       core.Blob
+	pfbIndex   int
+	blobIndex  int
+	shareSize  int
+	maxPadding int
+}
+
+func newElement(blob core.Blob, pfbIndex, blobIndex int) *element {
+	shareSize := shares.SparseSharesNeeded(uint32(len(blob.Data)))
+	return &element{
+		blob:      blob,
+		pfbIndex:  pfbIndex,
+		blobIndex: blobIndex,
+		shareSize: shareSize,
+		//
+		// For cacluating the maximum possible padding consider the following tree
+		// where each leaf corresponds to a share.
+		//
+		//	Depth       Position
+		//	0              0
+		//	              / \
+		//	             /   \
+		//	1           0     1
+		//	           /\     /\
+		//	2         0  1   2  3
+		//	         /\  /\ /\  /\
+		//	3       0 1 2 3 4 5 6 7
+		//
+		// Imagine if, according to the share commitment rules, a transcation took up 11 shares
+		// and had the merkle mountain tree commitment of 4,4,2,1. The first part of the share commitment
+		// would then be something that spans 4 shares and has a depth of 1. The worst case padding
+		// would be if the last transaction had a share at leaf index 0. Thus three padding shares would
+		// be needed to start the transaction at index 4 and be aligned with the first commitment.
+		// Thus the rule is to take the subtreewidh of the share size and subtract 1.
+		//
+		// Note that the padding would actually belong to the namespace of the transaction before it, but
+		// this makes no difference to the total share size.
+		maxPadding: shares.SubTreeWidth(shareSize) - 1,
```

Comment:
For the reviewer: Please make sure I have this assumption correct

Reply:
> Blobs start at an index that is equal to a multiple of the blob length divided by MaxSquareSize rounded up.

Where do you see this in ADR013? I can't find it

Reply:
Now we use `SubtreeRootThreshold` to work out the height in the tree of the highest merkle mountain and thus the width of the leaves

Reply:
> Where do you see this in ADR013? I can't find it

Sorry I should've linked directly to the line: https://github.com/celestiaorg/celestia-app/blob/f31ee58890b1f94815057b4bbd77d7cbe62172f4/docs/architecture/adr-013-non-interactive-default-rules-for-zero-padding.md?plain=1#L27